### PR TITLE
Add support for reloading the last checkpoint saved by passing `ckpt_path="last"`

### PR DESCRIPTION
**Pull request by otaj**
## What does this PR do?

Implements #11912 

Allows you to pass keyword `"last"` to `ckpt_path` argument. `"best"` still remains the default value if `None` is passed, except one particular case - in method `fit` when fault tolerant training is enabled.

Value `last` selects the checkpoint with the latest modified timestamp from set of all last checkpoints (currently only one, because they overwrite each other, but once #5030 is solved, it could be a different one) and all fault tolerant checkpoints. If no checkpoint in these two sets is found (due to fault tolerant training not being enabled and passing `save_last=False` to `ModelCheckpoint`), warning is issued and no checkpoint is loaded.

#### What needs to be done before making it non-draft

- [x] Write tests
- [x] Add changelog entry
- [x] Update documentation
    - [ ] is the change of docs big enough?
- [x] Discuss whether this design makes sense

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [x] Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or minor internal changes/refactors)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

cc @borda @awaelchli @ananthsub @ninginthecloud @rohitgr7 @carmocca @justusschock @jjenniferdai

### CODE REVIEW

File: pytorch_lightning/trainer/trainer.py [hunk 1]
```diff
@@ -1390,25 +1392,45 @@ def __set_ckpt_path(self, ckpt_path: Optional[str], model_provided: bool, model_
         if ft_checkpoints:
             ft_ckpt_path = ft_checkpoints[0].ckpt_path
             fs = get_filesystem(ft_ckpt_path)
-            if fs.exists(ft_ckpt_path):
-                return ft_ckpt_path
+            if not fs.exists(ft_ckpt_path):
+                ft_ckpt_path = None
+        else:
+            ft_ckpt_path = None
```
Comment (awaelchli) [inline pytorch_lightning/trainer/trainer.py:1398]: this can all be simplified by setting `ft_ckpt_path = None` from the beginning

File: pytorch_lightning/trainer/trainer.py [hunk 2]
```diff
@@ -1433,6 +1455,21 @@ def __set_ckpt_path(self, ckpt_path: Optional[str], model_provided: bool, model_
             # load best weights
             ckpt_path = self.checkpoint_callback.best_model_path
 
+        if ckpt_path == "last":
+            candidates = [ft.ckpt_path for ft in ft_checkpoints] + [
+                cb.last_model_path for cb in self.checkpoint_callbacks
+            ]
+            candidates_fs = {path: get_filesystem(path) for path in candidates if path}
+            candidates_ts = {path: fs.modified(path) for path, fs in candidates_fs.items() if fs.exists(path)}
```
Comment (awaelchli) [inline pytorch_lightning/trainer/trainer.py:1463]: I believe we should only take the last_model_path from the main checkpoint callback (the first one).

File: pytorch_lightning/trainer/trainer.py [hunk 3]
```diff
@@ -1390,25 +1392,45 @@ def __set_ckpt_path(self, ckpt_path: Optional[str], model_provided: bool, model_
...
 
         if model_provided and ckpt_path is None:
             # use passed model to function without loading weights
             return
 
-        fn = self.state.fn.value
-
         if model_connected and ckpt_path is None:
-            rank_zero_warn(
+            full_msg = (
```
Comment (awaelchli) [inline pytorch_lightning/trainer/trainer.py:1417]: can we clean this up a little? either we use f-string formatting only, or we do the string.format but mixing both is a bit harder to read and I believe it's not necessary.

File: tests/trainer/test_trainer.py [hunk 4]
```diff
@@ -664,6 +665,118 @@ def test_benchmark_option(benchmark_, deterministic, expected):
     torch.backends.cudnn.benchmark = original_val
 
 
+@pytest.mark.parametrize("ckpt_path", (None, "last"))
+@pytest.mark.parametrize("save_last", (True, False))
+@pytest.mark.parametrize("fn", ("fit", "validate"))
+@pytest.mark.parametrize("ft_enabled", (True, False))
+def test_checkpoint_path_input_last(tmpdir, ckpt_path, save_last, fn, ft_enabled):
```
Comment (awaelchli) [inline tests/trainer/test_trainer.py:672]: there is a bit much going on in this test. The highly nested if-else blocks make the test hard to follow and make the parametrizations almost pointless. 
I would at least disentangle the the fault tolerance from this test.

**Review by awaelchli — CHANGES_REQUESTED** [review]

File: pytorch_lightning/trainer/trainer.py (hunk 1 above)
Comment (otaj) [reply pytorch_lightning/trainer/trainer.py]: Done

File: pytorch_lightning/trainer/trainer.py (hunk 3 above)
Comment (otaj) [reply pytorch_lightning/trainer/trainer.py]: Done

File: tests/trainer/test_trainer.py (hunk 4 above)
Comment (otaj) [reply tests/trainer/test_trainer.py]: Cleaned up - now we have separate test for fault tolerancy enabled

File: pytorch_lightning/trainer/trainer.py (hunk 2 above)
Comment (otaj) [reply pytorch_lightning/trainer/trainer.py:1468]: We were talking with @carmocca about this and the conclusion was that at this moment, it does not really matter as they all point to the same file and in the future, once issue #5030 is done, then it will actually make sense the load the latest available from multiple runs. Does it make sense to you as well?

File: CHANGELOG.md [hunk 5]
```diff
@@ -9,6 +9,8 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/).
 
 ### Added
 
+- Allow to load last checkpoint using `ckpt_path="last"` ([#12816](https://github.com/PyTorchLightning/pytorch-lightning/pull/12816))
```
Comment (carmocca) [inline CHANGELOG.md:12]:
```suggestion
- Added support for reloading the last checkpoint saved by passing `ckpt_path="last"` ([#12816](https://github.com/PyTorchLightning/pytorch-lightning/pull/12816))
```

File: pytorch_lightning/trainer/trainer.py (hunk 2 above)
Comment (carmocca) [reply pytorch_lightning/trainer/trainer.py:1452]: Yes. Also note that when multiple checkpoint instances are used, not all might have `save_last=True` set

**Review by carmocca — APPROVED** [review]
LGTM! I've pushed some minor simplification to the code and greater simplifications to the test to narrow their scope to just these warnings.

File: tests/trainer/test_trainer.py [hunk 6]
```diff
@@ -666,71 +667,29 @@ def test_benchmark_option(benchmark_, deterministic, expected):
...
-    )
-    assert trainer.ckpt_path is None
-    trainer_fn = getattr(trainer, fn)
-
-    from pytorch_lightning.callbacks.fault_tolerance import _FaultToleranceCheckpoint
-
-    ft_checkpoints = [cb for cb in trainer.callbacks if isinstance(cb, _FaultToleranceCheckpoint)]
-    ft_ckpt_path = ft_checkpoints[0].ckpt_path
+@pytest.mark.parametrize("fn", (TrainerFn.FITTING, TrainerFn.VALIDATING))
+def test_checkpoint_path_input_last_fault_tolerant(ckpt_path, fn):
```
Comment (otaj) [inline tests/trainer/test_trainer.py:671]: This rewritten test now fails. The reason for that is that loading of last checkpoint checks whether the file exists and the fault tolerant checkpoint does not exist

Comment (otaj) [reply tests/trainer/test_trainer.py]: I fixed it with a simple addition

**Review by rohitgr7 — APPROVED** [review]
great!

**Review by kaushikb11 — APPROVED** [review]
